pkg/console/controllers/oauthclients/oauthclients.go

```diff
@@ -0,0 +1,203 @@
+		consoleOperatorLister: consoleOperatorInformer.Lister(),
+		routesLister:          routeInformer.Lister(),
+		ingressConfigLister:   ingressConfigInformer.Lister(),
+		targetNSSecretsLister: targetNSsecretsInformer.Lister(),
+
+		statusHandler: status.NewStatusHandler(operatorClient),
+	}
+
+	return factory.New().
+		WithSync(c.sync).
+		WithInformers(
+			oauthClientInformer.Informer(),
+			consoleOperatorInformer.Informer(),
+			routeInformer.Informer(),
+			ingressConfigInformer.Informer(),
+			targetNSsecretsInformer.Informer(),
+		).
+		WithSyncDegradedOnError(operatorClient).
+		ToController("OAuthClientsController", recorder.WithComponentSuffix("oauth-clients-controller"))
+}
+
+func (c *oauthClientsController) sync(ctx context.Context, controllerContext factory.SyncContext) error {
+	if shouldSync, err := c.handleStatus(ctx); err != nil {
+		return err
+	} else if !shouldSync {
+		return nil
+	}
+
+	operatorConfig, err := c.consoleOperatorLister.Get("cluster")
+	if err != nil {
+		return err
+	}
+
+	ingressConfig, err := c.ingressConfigLister.Get("cluster")
+	if err != nil {
+		return err
+	}
+
+	routeName := api.OpenShiftConsoleRouteName
+	routeConfig := routesub.NewRouteConfig(operatorConfig, ingressConfig, routeName)
+	if routeConfig.IsCustomHostnameSet() {
+		routeName = api.OpenshiftConsoleCustomRouteName
+	}
+
+	_, consoleURL, _, routeErr := operator.GetActiveRouteInfo(ctx, c.routesLister, routeName)
+	if routeErr != nil {
+		return routeErr
+	}
+
+	clientSecret, _, secErr := c.syncSecret(ctx, operatorConfig, controllerContext.Recorder())
+	c.statusHandler.AddConditions(status.HandleProgressingOrDegraded("OAuthClientSecretSync", "FailedApply", secErr))
+	if secErr != nil {
+		return c.statusHandler.FlushAndReturn(secErr)
+	}
+
+	oauthErrReason, oauthErr := c.syncOAuthClient(ctx, clientSecret, consoleURL.String())
+	c.statusHandler.AddConditions(status.HandleProgressingOrDegraded("OAuthClientSync", oauthErrReason, oauthErr))
+	if oauthErr != nil {
+		return c.statusHandler.FlushAndReturn(oauthErr)
+	}
+
+	return nil
+}
+
+// handleStatus returns whether sync should happen and any error encountering
+// determining the operator's management state
+func (c *oauthClientsController) handleStatus(ctx context.Context) (bool, error) {
+	operatorSpec, _, _, err := c.operatorClient.GetOperatorState()
+	if err != nil {
+		return false, fmt.Errorf("failed to retrieve operator config: %w", err)
+	}
+
+	switch managementState := operatorSpec.ManagementState; managementState {
+	case operatorv1.Managed:
+		klog.V(4).Infoln("console is in a managed state.")
+		return true, nil
+	case operatorv1.Unmanaged:
+		klog.V(4).Infoln("console is in an unmanaged state.")
+		return false, nil
+	case operatorv1.Removed:
+		klog.V(4).Infoln("console has been removed.")
+		return false, c.deregisterClient(ctx)
+	default:
+		return false, fmt.Errorf("console is in an unknown state: %v", managementState)
+	}
+}
```

Comment:
Could we add a TODO comment on this func to make it reusable for all of our controllers? We repeat this same logic in all of them.

Reply:
Absolutely. A long time ago I was thinking about how to add similar functionality to the controller factory itself but then there was a lot of other work and I never got back to it.